kayenta-signalfx/src/main/java/com/netflix/kayenta/signalfx/metrics/SignalFxMetricsService.java

```diff
@@ -0,0 +1,173 @@
+
+    // Fetch the user override or build the simple SignalFlow program.
+    String program = SimpleSignalFlowProgramBuilder
+        .create(queryConfig.getMetricName(), aggregationMethod)
+        .withQueryPairs(queryPairs)
+        .withScope(signalFxCanaryScope)
+        .build();
+
+    SignalFlowExecutionResult signalFlowExecutionResult;
+    try {
+      signalFlowExecutionResult = signalFlowService
+          .executeSignalFlowProgram(accessToken, startEpochMilli, endEpochMilli,
+              stepMilli, 0, true, program);
+    } catch (RetrofitError e) {
+      ErrorResponse errorResponse = (ErrorResponse) e.getBodyAs(ErrorResponse.class);
+      throw new SignalFxRequestError(errorResponse, program, startEpochMilli,
+          endEpochMilli, stepMilli, accountName);
+    }
+
+    // Return a Metric set of the reduced and aggregated data
+    MetricSet metricSet = MetricSet.builder()
+        .name(canaryMetricConfig.getName())
+        .startTimeMillis(startEpochMilli)
+        .startTimeIso(Instant.ofEpochMilli(startEpochMilli).toString())
+        .endTimeMillis(endEpochMilli)
+        .endTimeIso(Instant.ofEpochMilli(endEpochMilli).toString())
+        .stepMillis(stepMilli)
+        .values(getTimeSeriesDataFromChannelMessages(signalFlowExecutionResult.getChannelMessages()))
+        .tags(queryPairs.stream().collect(Collectors.toMap(QueryPair::getKey, QueryPair::getValue)))
+        .attribute("signal-flow-program", program)
+        .build();
+
+    return Collections.singletonList(metricSet);
+  }
+
+  /**
+   * Parses the data out of the SignalFx Signal Flow messages to build the data Kayenta needs to make judgements.
+   *
+   * @param channelMessages The list of messages from the signal flow execution.
+   * @return The list of values with missing data filled with NaNs
+   */
+  protected List<Double> getTimeSeriesDataFromChannelMessages(List<ChannelMessage> channelMessages) {
+    channelMessages.parallelStream().filter(channelMessage -> channelMessage.getType().equals(ERROR_MESSAGE))
+        .findAny()
+        .ifPresent(error -> {
+
+          // This error message is terrible, and I am not sure how to add more context to it.
+          // error.getErrors() returns a List<Object>, and it is unclear what to do with those.
+          throw new RuntimeException("Some sort of error occurred, when executing the signal flow program");
+        });
+
+    return channelMessages.parallelStream()
+        .filter(channelMessage -> channelMessage.getType().equals(DATA_MESSAGE))
+        .map(message -> {
+          ChannelMessage.DataMessage dataMessage = (ChannelMessage.DataMessage) message;
+          Map<String, Number> data = dataMessage.getData();
+          if (data.size() > 1) {
+            throw new IllegalStateException("There was more than one value for a given timestamp, a " +
+                "SignalFlow stream method that can aggregate, should have been applied to the data in " +
```

Comment:
I think you can drop the second comma from the exception message.

Reply:
- [ ] Fix grammer.